File: glut_opengl_estudos/10_sistema_particulas_blending.cpp
/*
Aula 10 — Sistema de Partículas com Blending Aditivo (C++)

Conceitos:
- Emissão, atualização e descarte de partículas por tempo de vida.
- Blending aditivo (`GL_SRC_ALPHA`, `GL_ONE`) para efeito de brilho.
- Organização orientada a dados para VFX simples.
*/

#include <GL/glut.h>
#include <cmath>
#include <random>
#include <vector>
#include "cpp_glut_utils.hpp"

namespace {

struct Particle {
    float x;
    float y;
    float vx;
    float vy;
    float life;
    float maxLife;
    float size;
};

class ParticleEmitter {
public:
    void update() {
        if (paused) return;

        for (int i = 0; i < emitRate; ++i) spawnOne();

        std::vector<Particle> alive;
        alive.reserve(particles.size());

        for (auto p : particles) {
            p.vy += gravity;
            p.x += p.vx;
            p.y += p.vy;
            p.life -= 0.016f;
            if (p.life > 0.0f) alive.push_back(p);
        }

        particles.swap(alive);
    }

    void render() const {
        glClear(GL_COLOR_BUFFER_BIT);
        glLoadIdentity();

        for (const auto& p : particles) {
            const float alpha = aula::clamp(p.life / p.maxLife, 0.0f, 1.0f);
            drawDisc(p.x, p.y, p.size, alpha);
        }

        glutSwapBuffers();
    }

    void clear() { particles.clear(); }

    int emitRate = 35;
    float gravity = -0.0007f;
    bool paused = false;

private:
    mutable std::mt19937 rng {7};
    std::vector<Particle> particles;

    float randomRange(float lo, float hi) {
        std::uniform_real_distribution<float> dist(lo, hi);
        return dist(rng);
    }

    void spawnOne() {
        const float angle = randomRange(-0.8f, 0.8f);
        const float speed = randomRange(0.008f, 0.018f);

        Particle p;
        p.x = 0.0f;
        p.y = -0.85f;
        p.vx = speed * angle;
        p.vy = randomRange(0.010f, 0.022f);
        p.life = randomRange(0.8f, 1.4f);
        p.maxLife = p.life;
        p.size = randomRange(0.010f, 0.025f);

        particles.push_back(p);
    }

    void drawDisc(float x, float y, float r, float alpha) const {
        glColor4f(1.0f, 0.55f, 0.12f, alpha);
        glBegin(GL_POLYGON);
        for (int i = 0; i < 20; ++i) {
            const float t = 2.0f * 3.1415926f * static_cast<float>(i) / 20.0f;
            glVertex2f(x + r * std::cos(t), y + r * std::sin(t));
        }
        glEnd();
    }
};

ParticleEmitter emitter;

void display() { emitter.render(); }

void keyboard(unsigned char key, int, int) {
    if (key == ' ') emitter.paused = !emitter.paused;
    else if (key == 'c' || key == 'C') emitter.clear();
    else if (key == '+') emitter.emitRate = aula::clamp(emitter.emitRate + 5, 5, 120);
    else if (key == '-') emitter.emitRate = aula::clamp(emitter.emitRate - 5, 5, 120);
    else if (key == 27) std::exit(0);
}

void tick(int) {
    emitter.update();
    glutPostRedisplay();
    glutTimerFunc(16, tick, 0);
}

} // namespace

int main(int argc, char** argv) {
    glutInit(&argc, argv);
    aula::configureCloseBehavior();
    glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB);
    glutInitWindowSize(1200, 850);
    glutCreateWindow("Aula 10 - Sistema de Particulas (C++)");

    glClearColor(0.02f, 0.02f, 0.05f, 1.0f);
    glEnable(GL_BLEND);
    glBlendFunc(GL_SRC_ALPHA, GL_ONE);

    glutDisplayFunc(display);
    glutKeyboardFunc(keyboard);
    glutTimerFunc(0, tick, 0);
    glutMainLoop();
    return 0;
}

File: glut_opengl_estudos/10_sistema_particulas_blending.py
"""Aula 10 — Sistema de partículas 2D com blending aditivo.

Exemplo avançado de VFX em OpenGL clássico com foco em arquitetura:
- Emissor configurável (taxa, espalhamento, gravidade, vida útil).
- Atualização estável de partículas com remoção eficiente.
- Render com blending aditivo para efeito de brilho.

Controles:
- ESPAÇO: pausa/retoma
- C: limpar partículas
- + / -: aumentar/reduzir taxa de emissão
- ESC: sair
"""

from __future__ import annotations

from dataclasses import dataclass
from random import Random

from OpenGL.GL import (
    GL_BLEND,
    GL_COLOR_BUFFER_BIT,
    GL_ONE,
    GL_ONE_MINUS_SRC_ALPHA,
    GL_POLYGON,
    GL_SRC_ALPHA,
    glBegin,
    glBlendFunc,
    glClear,
    glClearColor,
    glColor4f,
    glEnable,
    glEnd,
    glLoadIdentity,
    glRasterPos2f,
    glVertex2f,
)
from OpenGL.GLUT import (
    GLUT_ACTION_GLUTMAINLOOP_RETURNS,
    GLUT_ACTION_ON_WINDOW_CLOSE,
    GLUT_BITMAP_HELVETICA_12,
    GLUT_DOUBLE,
    GLUT_RGB,
    glutBitmapCharacter,
    glutCreateWindow,
    glutDisplayFunc,
    glutInit,
    glutInitDisplayMode,
    glutInitWindowSize,
    glutKeyboardFunc,
    glutMainLoop,
    glutPostRedisplay,
    glutSetOption,
    glutSwapBuffers,
    glutTimerFunc,
)


@dataclass
class Particle:
    """Partícula individual do sistema."""

    x: float
    y: float
    vx: float
    vy: float
    life: float
    max_life: float
    size: float


class ParticleEmitter:
    """Emissor de partículas com update e render desacoplados."""

    def __init__(self) -> None:
        self.rng = Random(7)
        self.particles: list[Particle] = []
        self.emit_rate = 35
        self.gravity = -0.0007
        self.paused = False

    def clear(self) -> None:
        """Remove todas as partículas ativas."""
        self.particles.clear()

    def _spawn_one(self) -> None:
        angle = self.rng.uniform(-0.8, 0.8)
        speed = self.rng.uniform(0.008, 0.018)
        vx = speed * angle
        vy = self.rng.uniform(0.010, 0.022)
        life = self.rng.uniform(0.8, 1.4)
        size = self.rng.uniform(0.010, 0.025)
        self.particles.append(Particle(0.0, -0.85, vx, vy, life, life, size))

    def update(self) -> None:
        """Atualiza estado físico e ciclo de vida das partículas."""
        if self.paused:
            return

        for _ in range(self.emit_rate):
            self._spawn_one()

        alive: list[Particle] = []
        for particle in self.particles:
            particle.vy += self.gravity
            particle.x += particle.vx
            particle.y += particle.vy
            particle.life -= 0.016
            if particle.life > 0.0:
                alive.append(particle)

        self.particles = alive

    def _draw_disc(self, x: float, y: float, radius: float, alpha: float) -> None:
        glColor4f(1.0, 0.55, 0.12, alpha)
        glBegin(GL_POLYGON)
        segments = 20
        from math import cos, sin, tau

        for i in range(segments):
            t = tau * i / segments
            glVertex2f(x + radius * cos(t), y + radius * sin(t))
        glEnd()

    def _draw_text(self, x: float, y: float, text: str) -> None:
        glColor4f(0.9, 0.9, 0.95, 1.0)
        glRasterPos2f(x, y)
        for char in text:
            glutBitmapCharacter(GLUT_BITMAP_HELVETICA_12, ord(char))

    def render(self) -> None:
        """Renderiza partículas com alpha baseado na vida restante."""
        glClear(GL_COLOR_BUFFER_BIT)
        glLoadIdentity()

        for particle in self.particles:
            alpha = max(0.0, min(1.0, particle.life / particle.max_life))
            self._draw_disc(particle.x, particle.y, particle.size, alpha)

        self._draw_text(-0.96, -0.94, f"particles={len(self.particles)} | rate={self.emit_rate} | paused={self.paused}")
        glutSwapBuffers()


emitter = ParticleEmitter()


def configure_close_behavior() -> None:
    try:
        glutSetOption(GLUT_ACTION_ON_WINDOW_CLOSE, GLUT_ACTION_GLUTMAINLOOP_RETURNS)
    except Exception:
        pass


def display() -> None:
    emitter.render()


def keyboard(key: bytes, _: int, __: int) -> None:
    if key == b" ":
        emitter.paused = not emitter.paused
    elif key in (b"c", b"C"):
        emitter.clear()
    elif key == b"+":
        emitter.emit_rate = min(emitter.emit_rate + 5, 120)
    elif key == b"-":
        emitter.emit_rate = max(emitter.emit_rate - 5, 5)
    elif key == b"\x1b":
        raise SystemExit("ESC")


def tick(_: int) -> None:
    emitter.update()
    glutPostRedisplay()
    glutTimerFunc(16, tick, 0)


def main() -> None:
    glutInit()
    configure_close_behavior()
    glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB)
    glutInitWindowSize(1200, 850)
    glutCreateWindow(b"Aula 10 - Sistema de Particulas com Blending")
    glClearColor(0.02, 0.02, 0.05, 1.0)

    glEnable(GL_BLEND)
    glBlendFunc(GL_SRC_ALPHA, GL_ONE)

    glutDisplayFunc(display)
    glutKeyboardFunc(keyboard)
    glutTimerFunc(0, tick, 0)

    print("Controles: ESPACO, C, +, -, ESC")
    glutMainLoop()


if __name__ == "__main__":
    main()
